api/src/main/java/com/spotify/ffwd/util/HighFrequencyDetector.java

```diff
@@ -0,0 +1,167 @@
+      new AtomicReference<>(new HashMap<>());
+
+  final AtomicLong highFrequencyTriggersTS;
+  private final Object lock = new Object();
+
+  /* TODO
+      1. unit tests
+          a. sendBatch with many points
+          b. send both with many points
+          c. check swap of high freq hashmap
+   */
+
+  @Inject
+  public HighFrequencyDetector() {
+    this.highFrequencyTriggersTS = new AtomicLong(System.currentTimeMillis());
+  }
+
+  /**
+   * Detects high frequency metrics by grouping and calculating
+   * time delta between metric timestamps
+   *
+   * @param metrics
+   * @return list of filtered metrics
+   */
+  public List<Metric> detect(final List<Metric> metrics) {
+    List<Metric> newList = new ArrayList<>();
+
+    // Groups metrics by metric hash code and finds times deltas of ordered data points.
+    // {hashcode -> -1, hashcode1 -> 10}
+    Map<Integer, Integer> groupedMetrics =
+        metrics.stream()
+            .sorted(Comparator.comparing(Metric::getTime))
+            .collect(
+                Collectors.groupingBy(
+                    Metric::hashCode,
+                    Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        this::computeTimeDelta)));
+
+    updateHighFrequencyMetricsStats(groupedMetrics);
+
+    if (dropHighFrequencyMetric && highFrequencyMetrics.get().size() > 0) {
+      metrics.stream()
+          .filter(
+              metric ->
+                  highFrequencyMetrics.get().getOrDefault(metric.hashCode(), 0)
+                      < minNumberOfTriggers)
+          .forEach(newList::add);
+
+      statistics.reportHighFrequencyMetricsDropped(metrics.size() - newList.size());
+      return newList;
+    }
+
+    return metrics;
+  }
+
+  private int computeTimeDelta(List<Metric> list) {
+    int size = list.size();
+    return IntStream.range(1, size)
+        .map(
+            x ->
+                (int)
+                    (list.get(size - x).getTime().getTime()
+                        - list.get(size - x - 1).getTime().getTime()))
+        .filter(d -> (d >= 0 && d < minFrequencyMillisAllowed))
+        .min()
+        .orElse(-1);
+  }
+
+  /**
+   * Updates internal map of high frequency metrics
+   *
+   * @param groupedMetrics - Grouped metrics
+   */
+  private void updateHighFrequencyMetricsStats(final Map<Integer, Integer> groupedMetrics) {
+    groupedMetrics.entrySet().stream()
+        .filter(x -> x.getValue() >= 0)
+        .forEach(
+            hashcode -> {
+              highFrequencyMetrics
+                  .get()
+                  .compute(hashcode.getKey(), (key, val) -> (val == null) ? 1 : val + 1);
+            });
+
+    statistics.reportHighFrequencyMetrics(highFrequencyMetrics.get().size());
+    swapHighFrequencyTriggersData();
+  }
+
+  /** Resets high frequency triggers data hashmap */
+  private void swapHighFrequencyTriggersData() {
+    synchronized (lock) {
```

Comment:
Ha I see what you are  trying to do with with the lock. You probably don't need a lock why not synchronize the entire method  and  use methods that provide atomicity for multiple operations for example  compareAndSet ...

Reply:
I think I can use `synchronized` method but not `compareAndSet` as I'm checking one atomic var and update two of them.